File: POTD/Matrix/Python/1914. Cyclically Rotating a Grid.py
class Solution:

    def updateGrid(self,sr:int,er:int,sc:int,ec:int,grid: List[List[int]], k: int):
        temp=[]

        # populate top row
        for j in range(sc,ec+1):
            temp.append(grid[sr][j])

        # popilaye last col
        for i in range(sr+1,er+1):
            temp.append(grid[i][ec])
        
        # populate last row
        for j in range(ec-1,sc-1,-1):
            temp.append(grid[er][j])
        
        # populate start row
        for i in range(er-1,sr,-1):
            temp.append(grid[i][sc])
        
        n=len(temp)
        k=k%n
        
        # populate back

        # populate top row
        rotIdx=0
        for j in range(sc,ec+1):
            putIdx=(rotIdx+k)%n
            grid[sr][j]=temp[putIdx]
            rotIdx += 1

        # popilaye last col
        for i in range(sr+1,er+1):
            putIdx=(rotIdx+k)%n
            grid[i][ec]=temp[putIdx]
            rotIdx+=1
        
        # populate last row
        for j in range(ec-1,sc-1,-1):
            putIdx=(rotIdx+k)%n
            grid[er][j]=temp[putIdx]
            rotIdx+=1
        
        # populate start row
        for i in range(er-1,sr,-1):
            putIdx=(rotIdx+k)%n
            grid[i][sc]=temp[putIdx]
            rotIdx+=1

    def rotateGrid(self, grid: List[List[int]], k: int) -> List[List[int]]:
        m=len(grid)
        n=len(grid[0])

        sr,er,sc,ec=0,m-1,0,n-1

        while sr<er and sc<ec:
            self.updateGrid(sr,er,sc,ec,grid,k)
            sr+=1
            er-=1
            sc+=1
            ec-=1
        
        return grid

File: POTD/Matrix/Cpp/1914. Cyclically Rotating a Grid.cpp
class Solution {
private:
void updateGrid(int sr,int er,int sc,int ec,vector<vector<int>>& grid, int k){
    vector<int> temp;
    // push first row
    for(int j=sc;j<=ec;j++){
        temp.push_back(grid[sr][j]);
    }
    // push last col
    for(int i=sr+1;i<=er;i++){
        temp.push_back(grid[i][ec]);
    }

    // push last row
    for(int j=ec-1;j>=sc;j--){
        temp.push_back(grid[er][j]);
    }

    // pushing starrting col
    for(int i=er-1;i>sr;i--){
        temp.push_back(grid[i][sc]);
    }

    int n=temp.size();
    vector<int> rotatedTemp(n);

    k=k%n;
    for(int i=0;i<n;i++){
        int rotIdx=(i-k+n)%n;
        rotatedTemp[rotIdx]=temp[i];
    }

    // now populate back into grid
    int rotIdx=0;

        // push first row
    for(int j=sc;j<=ec;j++){
        grid[sr][j]=rotatedTemp[rotIdx++];
    }
    // push last col
    for(int i=sr+1;i<=er;i++){
        grid[i][ec]=rotatedTemp[rotIdx++];
    }

    // push last row
    for(int j=ec-1;j>=sc;j--){
        grid[er][j]=rotatedTemp[rotIdx++];
    }

    // pushing starrting col
    for(int i=er-1;i>sr;i--){
        grid[i][sc]=rotatedTemp[rotIdx++];
    }
}
public:
    vector<vector<int>> rotateGrid(vector<vector<int>>& grid, int k) {
        int m=grid.size();
        int n=grid[0].size();

        int sr=0,er=m-1;
        int sc=0,ec=n-1;


        while(sr<er && sc<ec){
            updateGrid(sr,er,sc,ec,grid,k);
            sr++;
            er--;
            sc++;
            ec--;
        }

        return grid;
    }
};
